### PR TITLE
Arrow stream ingestion for JDBC client

### PR DESCRIPTION
**Pull request by hannes**
This adds a new method to the DuckDB JDBC client to read Arrow streams like views. Here is an example:

```Java
// arrow stuff
BufferAllocator allocator = new RootAllocator();
ArrowStreamReader reader = null; /* should not be null of course */
ArrowArrayStream arrow_array_stream = ArrowArrayStream.allocateNew(allocator);
Data.exportArrayStream(allocator, reader, arrow_array_stream);

// duckdb stuff
Class.forName("org.duckdb.DuckDBDriver");
DuckDBConnection conn = (DuckDBConnection) DriverManager.getConnection("jdbc:duckdb:");
conn.registerArrowStream("adsf", arrow_array_stream);

// run a query
Statement stmt = conn.createStatement();
ResultSet rs = (DuckDBResultSet) stmt.executeQuery("SELECT count(*) FROM adsf");
rs.next();
System.out.println(rs.getInt(1));
rs.close();

// clean up
stmt.close();
conn.close();
arrow_array_stream.close();
reader.close();
allocator.close();
```

CC @jonathanswenson

### CODE REVIEW

**Comment by pdet** [comment]
It looks about right.

I guess to do the pushdown we need to add a dependency? 

Not sure if that's already a solution on the works for it, but I'm guessing we can probably load the arrow jar at runtime with reflection? then decide at runtime if we want to push these or not?

File: tools/jdbc/src/jni/duckdb_java.cpp [hunk 1]
```diff
@@ -950,3 +951,54 @@ JNIEXPORT jlong JNICALL Java_org_duckdb_DuckDBNative_duckdb_1jdbc_1arrow_1stream
 	auto wrapper = new ResultArrowArrayStreamWrapper(move(res_ref->res), batch_size);
 	return (jlong)&wrapper->stream;
 }
+
+class JavaArrowTabularStreamFactory {
+public:
+	JavaArrowTabularStreamFactory(ArrowArrayStream *stream_ptr_p) : stream_ptr(stream_ptr_p) {};
+
+	~JavaArrowTabularStreamFactory() {
+		printf("close()\n");
```
Comment (jonathanswenson) [inline tools/jdbc/src/jni/duckdb_java.cpp:960]: nit: you have a print in here

**Comment by hannes** [comment]
@jonathanswenson can you please try this and report back?

**Comment by jonathanswenson** [comment]
@hannes Looks like this is working for me. Had to pull down the nightly arrow libraries (11.0.0-SNAPSHOT) from the nightlies to get it to work on my M1 mac, but was able to pull data in and query it directly, as well as do a CTAS statement to create a table. 

I ran into one slightly cryptic error when I tried to use a "registered" table that I had already closed the input stream for before running the query: 

```java.sql.SQLException: Invalid Input Error: arrow_scan: get_next failed(): java.lang.NullPointerException: Cannot invoke "java.nio.channels.ReadableByteChannel.read(java.nio.ByteBuffer)" because "this.in" is null
 	at org.apache.arrow.vector.ipc.ReadChannel.readFully(ReadChannel.java:60)
 	at org.apache.arrow.vector.ipc.message.MessageSerializer.readMessage(MessageSerializer.java:683)
 	at org.apache.arrow.vector.ipc.message.MessageChannelReader.readNext(MessageChannelReader.java:57)
 	at org.apache.arrow.vector.ipc.ArrowStreamReader.loadNextBatch(ArrowStreamReader.java:145)
 	at org.apache.arrow.c.ArrayStreamExporter$ExportedArrayStreamPrivateData.getNext(ArrayStreamExporter.java:72)
	at org.duckdb.DuckDBNative.duckdb_jdbc_execute(Native Method)
 	at org.duckdb.DuckDBPreparedStatement.execute(DuckDBPreparedStatement.java:131)
	at org.duckdb.DuckDBPreparedStatement.executeQuery(DuckDBPreparedStatement.java:154)
	at org.duckdb.DuckDBPreparedStatement.executeQuery(DuckDBPreparedStatement.java:182)
```

I created an ArrowStreamReader from a FileInputStream, Used that to register the name with duckdb, but closed the input stream before USING it in a query.

Not a problem, just had to decipher the error / figure out what I was doing wrong.

**Pull request merged**